src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs

```diff
@@ -271,6 +286,44 @@ private void CheckReachable(BoundStatement statement)
             }
         }
 
+        private void RecordTryFinallyReachability(BoundNode node)
+        {
+            Debug.Assert(node is BoundTryStatement
+                or BoundForEachStatement { AwaitOpt: not null }
+                or BoundUsingStatement { AwaitOpt: not null });
+
+            // If this is an async method, we need to track whether the end of the try block is reachable.
+            // This is used for async exception handler rewriting.
+            _asyncTryFinallyEndsReachable?.Add(node, State.Alive);
```

Comment:
From offline discussion, this will throw when adding the same node again, but that should be possible (looping around try/finally with a goto)